source/agora/common/Ensure.d

```diff
@@ -0,0 +1,60 @@
+/*******************************************************************************
+
+     Throw a statically allocated exception with formatting capacility.
+
+     The function in this module are not `@nogc` / `pure` but they do not
+     allocate if the formatting does not allocate.
+     A static buffer is used by the Exception, so the output length is limited.
+     Do not try to print a full block, for example.
+
+    Copyright:
+        Copyright (c) 2019-2021 BOSAGORA Foundation
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Ensure;
+
+import ocean.text.convert.Formatter;
+
+/// Module ctor to initialize the thread-local instance
+static this ()
+{
+    instance = new FormattedException();
+}
+
+/// The underlying instance
+private FormattedException instance;
+
+/// An exception which doesn't allocate but allows formatting a message
+private final class FormattedException : Exception
+{
+    /// Internal buffer, should be enough for most messages
+    private char[2048] buffer;
```

Comment:
What happens when message is too long? Would be good if it just truncates.

Reply:
Yes, that's what `snformat` does:
https://github.com/sociomantic-tsunami/ocean/blob/adb31c84baa2061d07aaa0cb7a7d14c3cc98309b/src/ocean/text/convert/Formatter.d#L182-L184